netutil/netutil.go

```diff
@@ -56,3 +59,166 @@ func privateNetworkInterfaces(all []net.Interface, fallback []string, logger log
 	level.Debug(logger).Log("msg", "found network interfaces with private IP addresses assigned", "interfaces", strings.Join(privInts, " "))
 	return privInts
 }
+
+// GetFirstAddressOf returns the first IPv4/IPV6 address of the supplied interface names, omitting any link-local addresses.
+func GetFirstAddressOf(names []string, logger log.Logger, preferInet6 bool) (string, error) {
+	return getFirstAddressOf(names, logger, getInterfaceAddresses, preferInet6)
+}
+
+// NetworkInterfaceGetter matches the signature of net.InterfaceByName() to allow for test mocks.
+type NetworkInterfaceAddressGetter func(name string) ([]netip.Addr, error)
+
+// getFirstAddressOf returns the first IPv4/IPV6 address of the supplied interface names, omitting any link-local addresses.
+func getFirstAddressOf(names []string, logger log.Logger, interfaceAddrsFunc NetworkInterfaceAddressGetter, preferInet6 bool) (string, error) {
+	var ipAddr netip.Addr
+
+	// When passing an empty list of interface names, we select all interfaces.
+	if len(names) == 0 {
+		infs, err := net.Interfaces()
+		if err != nil {
+			return "", fmt.Errorf("failed to get interface list and no interface names supplied: %w", err)
+		}
+		ifNames := make([]string, len(infs))
+		for i, v := range infs {
+			ifNames[i] = v.Name
+		}
+		names = ifNames
+	}
+
+	level.Debug(logger).Log("msg", "looking for addresses", "inf", fmt.Sprintf("%s", names), "inet6pref", preferInet6)
+
+	// Replace a nil func with the standard approach.
+	if interfaceAddrsFunc == nil {
+		interfaceAddrsFunc = getInterfaceAddresses
+	}
+
+	for _, name := range names {
+		addrs, err := interfaceAddrsFunc(name)
+		if err != nil {
+			level.Warn(logger).Log("msg", "error getting addresses for interface", "inf", name, "err", err)
+			continue
+		}
+
+		if len(addrs) <= 0 {
+			level.Warn(logger).Log("msg", "no addresses found for interface", "inf", name, "err", err)
+			continue
+		}
+		if ip := filterBestIP(addrs, preferInet6); ip.IsValid() {
+			// Select the best between what we've received
+			ipAddr = filterBestIP([]netip.Addr{ip, ipAddr}, preferInet6)
+		}
+		level.Debug(logger).Log("msg", "filtered best", "ipAddr", ipAddr.String(), "inf", name)
+		if ipAddr.IsLinkLocalUnicast() || !ipAddr.IsValid() {
+			level.Debug(logger).Log("msg", "skipping", "ipAddr", ipAddr.String(), "inf", name)
```

Comment:
This log was useful for developing, but perhaps we can just remove it.  What do you think?  For now I'll update to use "ignoring", since perhaps that is closer to what is happening.